File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.rb
def find_median_sorted_arrays(nums1, nums2)
  if nums1.length > nums2.length
    return find_median_sorted_arrays(nums2, nums1)
  end

  x = nums1.length
  y = nums2.length
  low = 0
  high = x

  while low <= high
    partition_x = (low + high) / 2
    partition_y = (x + y + 1) / 2 - partition_x

    max_left_x = (partition_x == 0) ? -Float::INFINITY : nums1[partition_x - 1]
    min_right_x = (partition_x == x) ? Float::INFINITY : nums1[partition_x]

    max_left_y = (partition_y == 0) ? -Float::INFINITY : nums2[partition_y - 1]
    min_right_y = (partition_y == y) ? Float::INFINITY : nums2[partition_y]

    if max_left_x <= min_right_y && max_left_y <= min_right_x
         if (x + y).even?
        return ([max_left_x, max_left_y].max + [min_right_x, min_right_y].min) / 2.0
      else
        return [max_left_x, max_left_y].max
      end
    elsif max_left_x > min_right_y
      # Move partition_x to the left
      high = partition_x - 1
    else
      low = partition_x + 1
    end
  end

  raise "Input arrays are not valid"
end

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size()) {
            // nums1'i her zaman daha kısa tut
            return findMedianSortedArrays(nums2, nums1);
        }

        int x = nums1.size();
        int y = nums2.size();
        int low = 0, high = x;

        while (low <= high) {
            int partitionX = (low + high) / 2;
            int partitionY = (x + y + 1) / 2 - partitionX;

            // Sol ve sağ bölmelerin elemanlarını belirle
            int maxLeftX = (partitionX == 0) ? INT_MIN : nums1[partitionX - 1];
            int minRightX = (partitionX == x) ? INT_MAX : nums1[partitionX];

            int maxLeftY = (partitionY == 0) ? INT_MIN : nums2[partitionY - 1];
            int minRightY = (partitionY == y) ? INT_MAX : nums2[partitionY];

            // Doğru bölmeyi bulduk mu?
            if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
                // Toplam eleman sayısı tek mi çift mi kontrol et
                if ((x + y) % 2 == 0) {
                    return (max(maxLeftX, maxLeftY) + min(minRightX, minRightY)) / 2.0;
                } else {
                    return max(maxLeftX, maxLeftY);
                }
            } 
            else if (maxLeftX > minRightY) {
                // Bölmeyi sola kaydır
                high = partitionX - 1;
            } 
            else {
                // Bölmeyi sağa kaydır
                low = partitionX + 1;
            }
        }

        throw invalid_argument("Input arrays are not valid.");
    }
};
